src/Dialect/ONNX/ONNXOps/ShapeHelper.cpp

```diff
@@ -378,49 +380,256 @@ bool ONNXBroadcastOpShapeHelper::hasNoBroadcast(DimAnalysis *dimAnalysis) {
+        return collapsedInnermostLoops > 0;
+      }
+    }
+
+    // 5) This dim is fine for manageable broadcasting. Account for the
+    // cumulative size of the inner dimensions.
+    // If all scalar, then dim is 1, and there is nothing to do. Otherwise
+    // accumulate in literal or dynamic sizes.
+    if (inputsDims[nonScalarID][rr].isLiteral()) {
+      collapsedLiteralSize *= inputsDims[nonScalarID][rr].getLiteral();
+    } else {
+      collapsedDynamicSize = collapsedDynamicSize * inputsDims[nonScalarID][rr];
+    }
+    collapsedInnermostLoops = -r;
+    LLVM_DEBUG(llvm::dbgs()
+               << "  SUCCESS at collapsing " << collapsedInnermostLoops
+               << " inner loops with cumulative static size of "
+               << collapsedLiteralSize << "\n\n");
+  } // For rank r.
+
+  // Came up to here, we are able to collapse them all.
+  return collapsedInnermostLoops > 0;
+}
+
 LogicalResult ONNXBroadcastOpShapeHelper::getAccessExprs(Value operand,
-    uint64_t operandIndex, const SmallVectorImpl<IndexExpr> &outputAccessExprs,
-    SmallVectorImpl<IndexExpr> &operandAccessExprs, bool hasNoBroadcast) {
+    int64_t operandIndex, const SmallVectorImpl<IndexExpr> &loopAccessExprs,
+    SmallVectorImpl<IndexExpr> &operandAccessExprs, bool flattenedInnerDims,
+    bool ruledOutBroadcast) {
+  // Get info.
+  int64_t loopDepth = loopAccessExprs.size();
+  int64_t inputSize = inputsDims.size();
+  int64_t operandRank = operand.getType().cast<ShapedType>().getRank();
+  // Flattened? no more than one loop per dim in output (aka output rank).
+  // Not flattened? one loop per dim in output (aka output rank).
+  if (flattenedInnerDims)
+    assert(loopDepth <= (int64_t)outputRank &&
+           "without flattening, expect one loop iter variable per output rank");
+  else
+    assert(loopDepth == (int64_t)outputRank &&
+           "without flattening, expect one loop iter variable per output rank");
+
   // Emtpy the access expr, just in case.
   operandAccessExprs.clear();
-  // There is this case where we have no broadcast per se, but we have mixtures
-  // of 1xTYPE vs TYPE scalars. Handle this case properly here.
-  uint64_t operandRank = operand.getType().cast<ShapedType>().getRank();
+  // There is this case where we have no broadcast per se, but we have
+  // mixtures of 1xTYPE vs TYPE scalars. Handle this case properly here.
   if (operandRank == 0)
     return success();
 
-  // The hasNoBroadcast pattern can be established by shape inference using
+  // The ruledOutBroadcast pattern can be established by shape inference using
   // DimAnalysis. If that is available, and broadcasting was ruled out, then
   // more efficient code can be generated.
-  if (hasNoBroadcast || (hasUniBroadcasting && operandIndex == 0)) {
-    for (IndexExpr ie : outputAccessExprs)
-      operandAccessExprs.emplace_back(ie);
-    return success();
-  }
+  bool noBroadcasting =
+      ruledOutBroadcast || (hasUniBroadcasting && operandIndex == 0);
```

Comment:
It looks to me `ruledOutBroadcast` is the same as the previous `hasNoBroadcast`, and the semantics of `hasNoBroadcast` is unchanged in this patch and still vague.

Reply:
Changed the names to be a bit clearer
```C++
  bool noBroadcasting =
      hasNoBroadcast || (hasUniBroadcasting && operandIndex == 0);
```